README.md
# Installation of Chroma and QUda 

## Into

## Usage

### Onestep

### Step by step

### Commands 
* gitclone 
* gitstatus
* install\_qmp
* install\_qio
* install\_qla
* install\_qdp
* install\_qopqdp
* install\_qdpxx
* install\_quda
* install\_chroma








NEWS
2020.11.23, Mon
- First version of My Makefile for installation of chroma and packages (qmp,qio,qla,qdp,qopqdp,qdp++);
  The cpu type works successfully on HPC@PKU

2021.05.07, Fri
- Add soft link symbol


refs/make_kelong.sh
#!/bin/bash -l
# Created by ybyang
# Modified by kelong

INSTALL=/public/home/kelong/software/chroma
INSTALL_PACKAGE=${INSTALL}/packages
precision="single"
# quda_path=/gpfs/share/home/1701110085/software/quda/quda-1.0.0/build

#flag='11000111' ## pure gpu
#flag='11111101' ## pure cpu
 flag='00000100' 

if [ ${flag:0:1} -eq '1' ]; then
dir="qmp";
# git clone  --recursive https://github.com/usqcd-software/${dir}.git
#mkdir build_${dir}
cd ${INSTALL_PACKAGE}/${dir}/build
#make clean
${INSTALL_PACKAGE}/${dir}/src/configure --prefix=${INSTALL_PACKAGE}/${dir}  \
     CXXFLAGS="-fopenmp -D_REENTRANT -g -O3 -finline-limit=50000 -std=c++11 -mavx2" \
     CC="mpicc" CXX="mpicxx"\
     CFLAGS="-fopenmp -D_REENTRANT -g -O3 -std=gnu99" --with-qmp-comms-type=MPI
make 
make install
cd ..
fi

if [ ${flag:1:1} -eq '1' ]; then
dir="qio";
# git clone --recursive  https://github.com/usqcd-software/${dir}.git
# git submodule update --init --recursive
# mkdir build_${dir}
cd ${INSTALL_PACKAGE}/${dir}/build
${INSTALL_PACKAGE}/${dir}/src/autogen.sh
${INSTALL_PACKAGE}/${dir}/src/configure --prefix=${INSTALL_PACKAGE}/${dir}  \
     CC="mpicc" CXX="mpicxx"\
     -with-qmp=${INSTALL}/build_qmp -enable-largefile -enable-dml-output-buffering\
     CFLAGS="-fopenmp -D_REENTRANT -g -O3 -std=gnu99" 

#make clean
make -j 4
make install
cd ..
fi

if [ ${flag:2:1} -eq '1' ]; then
dir="qla"
# git clone --recursive  https://github.com/usqcd-software/${dir}.git
# mkdir build_${dir}
cd ${INSTALL_PACKAGE}/${dir}/src
autoreconf -f
cd ${INSTALL_PACKAGE}/${dir}/build
# ${INSTALL_PACKAGE}/${dir}/src/autogen.sh
${INSTALL_PACKAGE}/${dir}/src/configure --prefix=${INSTALL_PACKAGE}/${dir}  \
     CC="mpicc" CXX="mpicxx"\
      --enable-sse3 --enable-temp-precision=D \
      CFLAGS="-Wall -std=gnu99 -O3 -mfpmath=sse -ffast-math -funroll-loops -fprefetch-loop-arrays -fomit-frame-pointer"
make -j
make install
cd ..
fi

if [ ${flag:3:1} -eq '1' ]; then
dir="qdp"
# git clone --recursive  https://github.com/usqcd-software/${dir}.git
mkdir build_${dir}
cd ${dir}
#make clean
autoreconf -f
./configure --prefix=${INSTALL}/build_${dir} \
     CC="mpicc" CXX="mpicxx"\
        --with-qmp=${INSTALL}/build_qmp --with-qio=${INSTALL}/build_qio --with-qla=${INSTALL}/qla --enable-sse\
        CFLAGS="-Wall -std=gnu99 -O3 -fargument-noalias-global -funroll-all-loops -fpeel-loops -ftree-vectorize"
make -j
make install
cd ..
fi

if [ ${flag:4:1} -eq '1' ]; then
dir="qopqdp"
# git clone --recursive  git://github.com/usqcd-software/${dir}.git
# mkdir build_${dir}
cd ${dir}
#make clean
autoreconf -f
./configure --prefix=${INSTALL_PACKAGE}/${dir} \
     CC="mpicc" CXX="mpicxx"\
        --with-qmp=${INSTALL_PACKAGE}/qmp --with-qio=${INSTALL_PACKAGE}/qio --with-qla=${INSTALL_PACKAGE}/qla --with-qdp=${INSTALL_PACKAGE}/qdp\
        --enable-underscores CFLAGS="-fopenmp -D_REENTRANT -g -std=gnu99 -O3"
make -j 16
make install
cd ..
fi

if [ ${flag:5:1} -eq '1' ]; then
dir="qdpxx"
# git clone --recursive  https://github.com/usqcd-software/${dir}.git
# git submodule update --init --recursive
# mkdir build_${dir}_${precision}
cd ${INSTALL_PACKAGE}/${dir}/${precision}/src
./autogen.sh
cd ${INSTALL_PACKAGE}/${dir}/${precision}/build
# ${INSTALL_PACKAGE}/${dir}/src/autogen.sh
${INSTALL_PACKAGE}/${dir}/${precision}/src/configure --prefix=${INSTALL_PACKAGE}/${dir}  \
     CC="mpicc" CXX="mpicxx"\
        --with-qmp=${INSTALL_PACKAGE}/qmp --enable-precision=${precision} --enable-parallel-arch=parscalar \
        --enable-largefile --enable-sse2 --enable-openmp \
        --enable-parallel-io --enable-dml-output-buffering \
        CFLAGS="-fopenmp -D_REENTRANT -g -std=gnu99 -O3 -mfpmath=sse -ffast-math -funroll-loops -fprefetch-loop-arrays -fomit-frame-pointer -ftree-vectorize -fassociative-math"\
        CXXFLAGS="-fopenmp -D_REENTRANT -g -std=c++11 -O3 -mfpmath=sse -ffast-math -funroll-loops -fprefetch-loop-arrays -fomit-frame-pointer -ftree-vectorize -fassociative-math"
# make  -j 8
# make install
cd ..
fi

if [ ${flag:6:1} -eq '1' ]; then
#for quda
make cmake
make hack
make build
#quda_path=../build_quda
fi

if [ ${flag:7:1} -eq '1' ]; then
dir="chroma"
BTYPE="cpu"
# git clone --recursive https://github.com/usqcd-software/${dir}.git
# git clone --recursive https://github.com/JeffersonLab/${dir}.git
#git submodule update --init --recursive
mkdir build_${dir}_${BTYPE}_${precision}
cd ${dir}
./autogen.sh
#make clean
./configure --prefix=${INSTALL2}/build_${dir}_${BTYPE}_${precision}
     FC=mpif90 CC="mpicc" CXX="mpicxx"\
        --with-qmp=${INSTALL}/build_qmp --with-qdp=${INSTALL}/build_qdpxx_${precision} --enable-qop-mg \
        --with-qio=${INSTALL}/build_qio --with-qla=${INSTALL}/build_qla --with-qdpc=${INSTALL}/build_qdp \
        --with-qopqdp=${INSTALL}/build_qopqdp --enable-precision=${precision} --enable-parallel-arch=parscalar\
        --enable-sse3 --enable-sse-wilson-dslash \
        CFLAGS="-fopenmp -D_REENTRANT -g -std=gnu99 -O3 -mfpmath=sse -ffast-math -funroll-loops -fprefetch-loop-arrays -fomit-frame-pointer -ftree-vectorize -fassociative-math"\
        CXXFLAGS="-fopenmp -D_REENTRANT -g -std=c++11 -O3 -mfpmath=sse -ffast-math -funroll-loops -fprefetch-loop-arrays -fomit-frame-pointer -ftree-vectorize -fassociative-math -fpermissive"
#        --with-quda=${quda_path}\
make
make install
fi


tmpbak/syssolver_linop_twisted_clover_quda_w.cc
/*! \file
 *  \brief Solve a MdagM*psi=chi linear system by CG2
 */

#include "chroma_config.h"
#ifdef BUILD_QUDA

#include "actions/ferm/invert/syssolver_linop_factory.h"
#include "actions/ferm/invert/syssolver_linop_aggregate.h"
#include "actions/ferm/invert/quda_solvers/syssolver_quda_clover_params.h"
#include "actions_ex/target_quda/syssolver_linop_twisted_clover_quda_w.h"
#include "io/aniso_io.h"


#include "handle.h"
#include "actions/ferm/fermstates/periodic_fermstate.h"
#include "actions/ferm/linop/lwldslash_w.h"
#include "meas/glue/mesplq.h"
// QUDA Headers
#include <quda.h>
// #include <util_quda.h>

namespace Chroma
{
  namespace LinOpSysSolverQUDATwistedCloverEnv
  {

    //! Anonymous namespace
    namespace
    {
      //! Name to be used
      const std::string name("QUDA_TWISTED_CLOVER_INVERTER");

      //! Local registration flag
      bool registered = false;
    }



    LinOpSystemSolver<LatticeFermion>* createFerm(XMLReader& xml_in,	
						  const std::string& path,
						  Handle< FermState< LatticeFermion, multi1d<LatticeColorMatrix>, multi1d<LatticeColorMatrix> > > state, 
						  
						  Handle< LinearOperator<LatticeFermion> > A)
    {
      return new LinOpSysSolverQUDATwistedClover(A, state,SysSolverQUDACloverParams(xml_in, path));
    }

    //! Register all the factories
    bool registerAll() 
    {
      bool success = true; 
      if (! registered)
      {
	success &= Chroma::TheLinOpFermSystemSolverFactory::Instance().registerObject(name, createFerm);
	registered = true;
      }
      return success;
    }
  }

  SystemSolverResults_t 
#ifdef _PDF_
  LinOpSysSolverQUDATwistedClover::qudaInvert(const CloverTermT<T, U>::Type_t& clover,
				       const CloverTermT<T, U>::Type_t& invclov,
				       const T& chi_s,
				       T& psi_s) const{
#else				       
  LinOpSysSolverQUDATwistedClover::qudaInvert(const CloverTermT<T, U>& clover,
				       const CloverTermT<T, U>& invclov,
				       const T& chi_s,
				       T& psi_s) const{
#endif				       

    SystemSolverResults_t ret;

    void *spinorIn;

    T mod_chi;
    if ( quda_inv_param.matpc_type == QUDA_MATPC_ODD_ODD_ASYMMETRIC ) {
      // asymmetric 
      //
      // Solve A_oo - D A^{-1}_ee D -- chroma conventions.
      // No need to transform source
#ifndef BUILD_QUDA_DEVIFACE_SPINOR
      spinorIn =(void *)&(chi_s.elem(rb[1].start()).elem(0).elem(0).real());
#else
      spinorIn = QDPCache::Instance().getDevicePtr( chi_s.getId() );
      QDPIO::cout << "MDAGM spinor in = " << spinorIn << "\n";
#endif
    }
    else if( quda_inv_param.matpc_type == QUDA_MATPC_ODD_ODD) { 
      //
      // symmetric
      // Solve with M_symm = 1 - A^{-1}_oo D A^{-1}ee D 
      //
      // Chroma M =  A_oo ( M_symm )
      //
      //  So  M x = b => A_oo (M_symm) x = b 
      //              =>       M_symm x = A^{-1}_oo b = chi_mod
      invclov.apply(mod_chi, chi_s, PLUS, 1);
#ifndef BUILD_QUDA_DEVIFACE_SPINOR
      spinorIn =(void *)&(mod_chi.elem(rb[1].start()).elem(0).elem(0).real());
#else
      spinorIn = QDPCache::Instance().getDevicePtr( mod_chi.getId() );
      QDPIO::cout << "MDAGM spinor in = " << spinorIn << "\n";
#endif
    }
    else { 
      QDPIO::cout << "MATPC Type not allowed." << std::endl;
      QDPIO::cout << " Allowed are: QUDA_MATPC_ODD_ODD_ASYMMETRIC or QUDA_MATPC_ODD_ODD" << std::endl;
      QDP_abort(1);
    }

#ifndef BUILD_QUDA_DEVIFACE_SPINOR
    void* spinorOut =(void *)&(psi_s.elem(rb[1].start()).elem(0).elem(0).real());
#else
    void* spinorOut = QDPCache::Instance().getDevicePtr( psi_s.getId() );
    QDPIO::cout << "MDAGM spinor out = " << spinorOut << "\n";
#endif

    // Do the solve here 
    StopWatch swatch1; 
    swatch1.reset();
    swatch1.start();
    invertQuda(spinorOut, spinorIn, (QudaInvertParam*)&quda_inv_param);
    swatch1.stop();


    QDPIO::cout << "Cuda Space Required" << std::endl;
//    QDPIO::cout << "\t Spinor:" << quda_inv_param.spinorGiB << " GiB" << std::endl;
//    QDPIO::cout << "\t Gauge :" << q_gauge_param.gaugeGiB << " GiB" << std::endl;
//    QDPIO::cout << "\t InvClover :" << quda_inv_param.cloverGiB << " GiB" << std::endl;
    QDPIO::cout << "QUDA_"<<solver_string<<"_CLOVER_SOLVER: time="<< quda_inv_param.secs <<" s" ;
    QDPIO::cout << "\tPerformance="<<  quda_inv_param.gflops/quda_inv_param.secs<<" GFLOPS" ; 
    QDPIO::cout << "\tTotal Time (incl. load gauge)=" << swatch1.getTimeInSeconds() <<" s"<<std::endl;

    ret.n_count =quda_inv_param.iter;

    return ret;

  }
  

}


 
// DEAD Test Code
#if 0
    void* spinorIn =(void *)&(mod_chi.elem(rb[1].start()).elem(0).elem(0).real());
    void* spinorOut =(void *)&(psi_s.elem(rb[0].start()).elem(0).elem(0).real());
    
    // OK Here I have a chance to test directly 
    // Even Target Checkerboard, No Dagger
    psi_s = zero;
    quda_inv_param.dslash_type = QUDA_TWISTED_CLOVER_DSLASH; // Sets Clover Matrix    
    dslashQuda(spinorOut, spinorIn, &quda_inv_param, 0, 0);



    // Need to create a simple ferm state from the links_single...
    Handle< FermState<T, Q, Q> > pstate(new PeriodicFermState<TF,QF,QF>(links_orig));
    const AnisoParam_t& aniso = invParam.CloverParams.anisoParam;
    QDPWilsonDslashT<T,Q,Q>  qdp_dslash(pstate, aniso);
    
    T tmp,psi2;
    tmp=zero;
    psi2=zero;
    // qdp_dslash.apply(psi2, mod_chi, PLUS, 0);
    qdp_dslash.apply(tmp, mod_chi, PLUS, 0);
    invclov.apply(psi2,tmp, PLUS, 0);


    T r=zero;
    r = psi2 - psi_s;
    
    QDPIO::cout << "CB=0" << std::endl;
    QDPIO::cout << "Dslash Test: || r || = " << sqrt(norm2(r,rb[0])) << std::endl;
    //    QDPIO::cout << "Dslash Test: || r ||/|| psi || = " << sqrt(norm2(r,rb[0])/norm2(psi_s, rb[0])) << std::endl;

    QDPIO::cout << "CB=1: Should be zero" << std::endl;
    QDPIO::cout << "Dslash Test: || r || = " << sqrt(norm2(r,rb[1])) << std::endl;
    //QDPIO::cout << "Dslash Test: || r ||/|| psi || = " << sqrt(norm2(r,rb[1])/norm2(psi_s, rb[1])) << std::endl;
    
    const int* tab = rb[0].siteTable().slice();
    for(int i=0; i < rb[0].numSiteTable(); i++) { 
      int j = tab[i];
      bool printSite=false;

      for(int spin=0; spin < 4; spin++) {
	for(int col=0; col < 3; col++) { 
	  if( (fabs(r.elem(j).elem(spin).elem(col).real()) > 1.0e-5 )
	      || (fabs(r.elem(j).elem(spin).elem(col).imag()) > 1.0e-5 )) {
	    printSite=true;
	  }
	}
      }
      if( printSite ) { 
	  
	for(int spin=0; spin < 4; spin++) { 
	  for(int col=0; col < 3; col++) { 
	    QDPIO::cout << "Site= " << j << " Spin= "<< spin << " Col= " << col << " spinor = ( " 
			<< psi2.elem(j).elem(spin).elem(col).real()  << " , " 
			<< psi2.elem(j).elem(spin).elem(col).imag()  << " )" << std::endl;
	  }
	}
	QDPIO::cout << std::endl;
      }
    }
    QDP_abort(1);
#endif
#endif


refs/make.sh
#!/bin/bash -l

INSTALL=/public/home/ybyang/opt/new
INSTALL2=/public/home/ybyang/opt/new
precision="double"
quda_path=/public/home/ybyang/soft/build_quda_new

#flag='11000111' ## pure gpu
#flag='11111101' ## pure cpu
flag='00000002' 

if [ ${flag:0:1} -eq '1' ]; then
dir="qmp";
#git clone --recursive  git://github.com/usqcd-software/${dir}.git
mkdir build_${dir}
cd build_${dir}
#make clean
../${dir}/configure --prefix=${INSTALL}/${dir}  \
     CC="mpicc" CXX="mpicxx"\
     CXXFLAGS="-fopenmp -D_REENTRANT -g -O3 -finline-limit=50000 -std=c++11 -mavx2" \
     CFLAGS="-fopenmp -D_REENTRANT -g -O3 -std=gnu99" --with-qmp-comms-type=MPI
make -j
make install
cd ..
fi

if [ ${flag:1:1} -eq '1' ]; then
dir="qio";
#git clone --recursive  git://github.com/usqcd-software/${dir}.git
mkdir build_${dir}
cd build_${dir}
#make clean
#./autogen.sh
../${dir}/configure --prefix=${INSTALL}/${dir} \
     CC="mpicc" CXX="mpicxx"\
     -with-qmp=${INSTALL}/qmp -enable-largefile -enable-dml-output-buffering\
     CFLAGS="-fopenmp -D_REENTRANT -g -O3 -std=gnu99" 
make -j
make install
cd ..
fi

if [ ${flag:2:1} -eq '1' ]; then
dir="qla"
#git clone --recursive  git://github.com/usqcd-software/${dir}.git
mkdir build_${dir}
cd build_${dir}
#make clean
../${dir}/configure --prefix=${INSTALL}/${dir} \
     CC="mpicc" CXX="mpicxx"\
      --enable-sse3 --enable-temp-precision=D \
      CFLAGS="-Wall -std=gnu99 -O3 -mfpmath=sse -ffast-math -funroll-loops -fprefetch-loop-arrays -fomit-frame-pointer"
make -j
make install
cd ..
fi

if [ ${flag:3:1} -eq '1' ]; then
dir="qdp"
#git clone --recursive  git://github.com/usqcd-software/${dir}.git
mkdir build_${dir}
cd build_${dir}
#make clean
../${dir}/configure --prefix=${INSTALL}/${dir} \
     CC="mpicc" CXX="mpicxx"\
        --with-qmp=${INSTALL}/qmp --with-qio=${INSTALL}/qio --with-qla=${INSTALL}/qla --enable-sse\
        CFLAGS="-Wall -std=gnu99 -O3 -fargument-noalias-global -funroll-all-loops -fpeel-loops -ftree-vectorize"
make -j
make install
cd ..
fi

if [ ${flag:4:1} -eq '1' ]; then
dir="qopqdp"
#git clone --recursive  git://github.com/usqcd-software/${dir}.git
mkdir build_${dir}
cd build_${dir}
#make clean
../${dir}/configure --prefix=${INSTALL}/${dir} \
     CC="mpicc" CXX="mpicxx"\
        --with-qmp=${INSTALL}/qmp --with-qio=${INSTALL}/qio --with-qla=${INSTALL}/qla --with-qdp=${INSTALL}/qdp\
        --enable-underscores CFLAGS="-fopenmp -D_REENTRANT -g -std=gnu99 -O3"
make -j
make install
cd ..
fi

if [ ${flag:5:1} -eq '1' ]; then
dir="qdpxx"
#git clone --recursive  git://github.com/usqcd-software/${dir}.git
mkdir build_${dir}_${precision}
cd build_${dir}_${precision}
../${dir}/configure --prefix=${INSTALL}/${dir}_${precision}\
     CC="mpicc" CXX="mpicxx"\
        --with-qmp=${INSTALL}/qmp --enable-precision=${precision} --enable-parallel-arch=parscalar\
        --enable-largefile --enable-sse2 --enable-openmp\
        --enable-parallel-io --enable-dml-output-buffering \
        CFLAGS="-fopenmp -D_REENTRANT -g -std=gnu99 -O3 -mfpmath=sse -ffast-math -funroll-loops -fprefetch-loop-arrays -fomit-frame-pointer -ftree-vectorize -fassociative-math"\
        CXXFLAGS="-fopenmp -D_REENTRANT -g -std=c++11 -O3 -mfpmath=sse -ffast-math -funroll-loops -fprefetch-loop-arrays -fomit-frame-pointer -ftree-vectorize -fassociative-math"
make  -j
make install
cd ..
fi

if [ ${flag:6:1} -eq '1' ]; then
#for quda
make cmake
make hack
make build
#quda_path=../build_quda
fi

if [ ${flag:7:1} -eq '1' ]; then
dir="chroma"
mkdir build_${dir}_${precision}_cpu
cd build_${dir}_${precision}_cpu
#make clean
../${dir}/configure --prefix=${INSTALL2}/${dir}_cpu_${precision} \
     FC=mpif90 CC="mpicc" CXX="mpicxx"\
        --with-qmp=${INSTALL}/qmp --with-qdp=${INSTALL}/qdpxx_${precision} --enable-qop-mg \
        --with-qio=${INSTALL}/qio --with-qla=${INSTALL}/qla --with-qdpc=${INSTALL}/qdp \
        --with-qopqdp=${INSTALL}/qopqdp --enable-precision=${precision} --enable-parallel-arch=parscalar\
        --enable-sse3 --enable-sse-wilson-dslash \
        CFLAGS="-fopenmp -D_REENTRANT -g -std=gnu99 -O3 -mfpmath=sse -ffast-math -funroll-loops -fprefetch-loop-arrays -fomit-frame-pointer -ftree-vectorize -fassociative-math"\
        CXXFLAGS="-fopenmp -D_REENTRANT -g -std=c++11 -O3 -mfpmath=sse -ffast-math -funroll-loops -fprefetch-loop-arrays -fomit-frame-pointer -ftree-vectorize -fassociative-math -fpermissive"

make -j 8
make install
fi

if [ ${flag:7:1} -eq '2' ]; then
dir="chroma"
mkdir build_${dir}_${precision}
cd build_${dir}_${precision}
#make clean
../${dir}/configure --prefix=${INSTALL2}/${dir}_gpu_${precision} \
     FC=mpif90 CC="mpicc" CXX="mpicxx"\
        --with-qmp=${INSTALL}/qmp --with-qdp=${INSTALL}/qdpxx_${precision} --enable-qop-mg \
        --with-qio=${INSTALL}/qio --with-qla=${INSTALL}/qla --with-qdpc=${INSTALL}/qdp \
        --with-qopqdp=${INSTALL}/qopqdp --enable-precision=${precision} --enable-parallel-arch=parscalar\
        --enable-sse3 --enable-sse-wilson-dslash \
	--with-quda=${quda_path}\
        CFLAGS="-fopenmp -D_REENTRANT -g -std=gnu99 -O3 -mfpmath=sse -ffast-math -funroll-loops -fprefetch-loop-arrays -fomit-frame-pointer -ftree-vectorize -fassociative-math"\
        CXXFLAGS="-fopenmp -D_REENTRANT -g -std=c++11 -O3 -mfpmath=sse -ffast-math -funroll-loops -fprefetch-loop-arrays -fomit-frame-pointer -ftree-vectorize -fassociative-math -fpermissive"
make -j 8
make install
fi
